File: leetcode/java/0501_find_mode_in_binary_search_tree.java
/**
 * 0501. Find Mode in Binary Search Tree
 * Problem Link: https://leetcode.com/problems/find-mode-in-binary-search-tree/
 * Solution Link (Personal):
 */

class Solution {
  Integer prev = null;
  int count = 1;
  int max = 0;

  public int[] findMode(TreeNode root) {
      if (root == null) return new int[0];

      List<Integer> modes = new ArrayList<>();
      traverse(root, modes);

      int[] result = new int[modes.size()];
      for (int i = 0; i < modes.size(); ++i) {
          result[i] = modes.get(i);
      }
      return result;
  }

  private void traverse(TreeNode node, List<Integer> modes) {
      if (node == null) return;

      traverse(node.left, modes);

      if (prev != null) {
          if (node.val == prev) {
              count++;
          } else {
              count = 1;
          }
      }

      if (count > max) {
          max = count;
          modes.clear();
          modes.add(node.val);
      } else if (count == max) {
          modes.add(node.val);
      }

      prev = node.val;

      traverse(node.right, modes);
  }
}

File: leetcode/python3/0501_find_mode_in_binary_search_tree.py
"""
 * 501. Find Mode in Binary Search Tree 
 * Problem Link: https://leetcode.com/problems/find-mode-in-binary-search-tree/
 * Solution Link (Personal):
"""

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def findMode(self, root):
        def inOrderTraversal(node):
            nonlocal current_val, current_count, max_count, modes
            if node:
                inOrderTraversal(node.left)
                if node.val == current_val:
                    current_count += 1
                else:
                    current_val = node.val
                    current_count = 1
                if current_count > max_count:
                    max_count = current_count
                    modes = [current_val]
                elif current_count == max_count:
                    modes.append(current_val)
                inOrderTraversal(node.right)

        if not root:
            return []

        current_val = None
        current_count = 0
        max_count = 0
        modes = []

        inOrderTraversal(root)

        return modes
